crucible-llvm/src/Lang/Crucible/LLVM/SymIO.hs

```diff
@@ -0,0 +1,560 @@
+           Just fileHandle -> \_ ->
+             SymIO.closeFileHandle (llvmFileSystem fsVars) fileHandle $ \case
+               Just SymIO.FileHandleClosed -> returnIOError32
+               Nothing -> liftIO $ W4.bvLit sym (PN.knownNat @32) (BVS.mkBV (PN.knownNat @32) 0)
+           Nothing -> \_ -> returnIOError32
+  )
+
+
+readFileHandle
+  :: (IsSymInterface sym, HasLLVMAnn sym, HasPtrWidth wptr)
+  => LLVMFileSystem wptr
+  -> LLVMOverride p sym
+           (EmptyCtx ::> BVType 32
+                     ::> LLVMPointerType wptr
+                     ::> BVType wptr)
+           (BVType wptr)
+readFileHandle fsVars =
+  [llvmOvr| ssize_t @read( i32, i8*, size_t ) |]
+  (\memOps sym args@(Empty :> filedesc :> _ :> _ ) ->
+       do
+         lookupFileHandle fsVars (regValue filedesc) (RegMap args) $ \case
+           Just fileHandle -> \(RegMap (Empty :> _ :> buffer_ptr :> size)) ->
+             SymIO.readChunk (llvmFileSystem fsVars) fileHandle (regValue size) $ \case
+               Left SymIO.FileHandleClosed -> returnIOError
+               Right (chunk, bytesRead) -> do
+                 modifyGlobal memOps $ \mem -> liftIO $ do
+                   chunkArray <- SymIO.chunkToArray sym (W4.BaseBVRepr PtrWidth) chunk
+                   mem' <- doArrayStore sym mem (regValue buffer_ptr) noAlignment chunkArray bytesRead
+                   return (bytesRead, mem')
+           Nothing -> \_ -> returnIOError
+  )
+
+-- | If the write is to a concrete FD for which we have an associated 'IO.Handle', mirror the write to that Handle
+--
+-- This is intended to support mirroring stdout/stderr in a user-visible way
+-- (noting that symbolic IO can be repeated due to the simulator branching).
+-- Note also that only writes of a concrete length are mirrored reasonably;
+-- writes with symbolic length are denoted with a substitute token. Likewise
+-- individual symbolic bytes are printed as @?@ characters.
+doConcreteWrite
+  :: (IsSymInterface sym, HasPtrWidth wptr)
+  => PN.NatRepr wptr
+  -> Map.Map Natural IO.Handle
+  -> RegValue sym (BVType 32)
+  -> SymIO.DataChunk sym wptr
+  -> RegEntry sym (BVType wptr)
+  -> OverrideSim p sym ext rtp args ret ()
+doConcreteWrite ptrw handles symFD chunk size =
+  case W4.asBV symFD of
+    Just (BVS.asNatural -> fd)
+      | Just hdl <- Map.lookup fd handles
+      , Just numBytes <- BVS.asUnsigned <$> W4.asBV (regValue size) -> do
+          -- We have a concrete size of the write and an IO Handle to write
+          -- to. Write each byte that is concrete, or replace it with a '?'
+          sym <- getSymInterface
+          F.forM_ [0..numBytes - 1] $ \idx -> do
+            idxBV <- liftIO $ W4.bvLit sym ptrw (BVS.mkBV ptrw idx)
+            byteVal <- liftIO $ SymIO.evalChunk chunk idxBV
+            case W4.asBV byteVal of
+              Just (BVS.asUnsigned -> concByte) -> liftIO $ BS.hPut hdl (BS.pack [fromIntegral concByte])
+              Nothing -> liftIO $ BS.hPut hdl (BS.pack [63])
```

Comment:
Is 63 the ASCII code for '?' ... let's comment that.

Reply:
Yeah. I was being too lazy. Let me just import Data.ByteString.Char8 and do the right thing here. Thanks for the reminder

Reply:
Okay, fixed in f9cfe51